test/suite/wthooks.py

```diff
@@ -0,0 +1,256 @@
+                    self.hooks.append(hook)
+            except:
+                print('Cannot import hook: ' + name + ', check file ' + modname + '.py')
+                raise
+        for hook in self.hooks:
+            hook.setup_hooks()
+
+    def add_hook(self, clazz, method_name, hook_type, hook_func):
+        if not hasattr(clazz, method_name):
+            raise Exception('Cannot find method ' + method_name + ' on class ' + str(clazz))
+
+        # We need to set up some extra attributes on the Connection class.
+        # Given that the method name is XXXX, and class is Connection, here's what we're doing:
+        #    wiredtiger.Connection._XXXX_hooks = []
+        #    wiredtiger.Connection._XXXX_orig = wiredtiger.Connection.XXXX
+        #    wiredtiger.Connection.XXXX = lambda self, *args: hooked_function(self, 'XXXX', *args)
+        hooks_name = '_' + method_name + '_hooks'
+        orig_name = '_' + method_name + '_orig'
+        if not hasattr(clazz, hooks_name):
+            #tty('Setting up hook on ' + str(clazz) + '.' + method_name)
+            setattr(clazz, hooks_name, [])
+            orig_func = getattr(clazz, method_name)
+            if orig_func == None:
+                raise Exception('method ' + method_name + ' hook setup: method does not exist')
+
+            # If we're using the wiredtiger module and not a class, we need a slightly different
+            # style of hooked_function, since there is no self.  What would be the "self" argument
+            # is in fact the class.
+            if clazz == wiredtiger:
+                f = lambda *args: hooked_function(wiredtiger, orig_func, hooks_name, *args)
+            else:
+                f = lambda self, *args: hooked_function(self, orig_func, hooks_name, *args)
+            setattr(clazz, method_name, f)
+            setattr(clazz, orig_name, orig_func)
+
+        # Now add to the list of hook functions
+        # If it's a replace hook, we only allow one of them for a given method name
+        hooks_list = getattr(clazz, hooks_name)
+        if hook_type == HOOK_REPLACE:
+            for ht, hf in hooks_list:
+                if ht == HOOK_REPLACE:
+                    raise Exception('method ' + method_name + ' hook setup: trying to replace the same method with two hooks')
+        elif hook_type != HOOK_NOTIFY and hook_type != HOOK_ARGS:
+            raise Exception('method ' + method_name + ' hook setup: unknown hook_type: ' + str(hook_type))
+        hooks_list.append([hook_type, hook_func])
+        setattr(clazz, hooks_name, hooks_list)
+        #tty('Setting up hooks list in ' + str(clazz) + '.' + hooks_name)
+
+    def get_function(self, clazz, method_name):
+        orig_name = '_' + method_name + '_orig'
+        if hasattr(clazz, orig_name):
+            orig_func = getattr(clazz, orig_name)
+        else:
+            orig_func = getattr(clazz, method_name)
+        return orig_func
+
+    def filter_tests(self, tests):
```

Comment:
Does this imply that multiple hooks will always create a smaller or equal test suite? What happens if you want a hook to specifically run under a specific test suite, and another hook to run another different test suite? That would mean no tests would be run, let me know if that's the intended use case

Reply:
Yes, multiple hooks will run a smaller or equal test suite. It's a very good point.  I think this filter was less intended to be used as "only run the cursor tests with this hook", but more like, "don't run the backup tests with this hook, there's something fundamentally broken about that combination".   Similar to how we "skip" tests today, quite targeted.  So, yes it's possible that combining multiple hooks might get you to empty set, I think the filters should be cautious about removing too much.  Especially if we have multiple hooks in mind.  For at least a while, I'm suspecting that multiple hooking will either be rare or else someone may create a few hooks at the same time specially expecting them to be used together.